### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by creator23**
Implemented Assignment 2- A simple web server, and config parser. The webserver just echos the request it receives.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+CXX=g++
+CXXOPTIMIZE= -O2
+CXXFLAGS= -g -Wall -pthread -std=c++0x $(CXXOPTIMIZE)
```
Comment (creator23) [inline Makefile]: c++0x is out of date, use c++11

File: server.cpp [hunk 2]
```diff
@@ -34,10 +34,18 @@ void connection::do_write() {
 	auto self(shared_from_this());
 
 	reply_.status = reply::ok;
-	reply_.headers[0].name = "Content-Length";
-	reply_.headers[0].value= std::to_string(reply_.content.size());
-	reply_.headers[1].name = "Content-Type";
-	reply_.headers[1].value = "text/plain";
+  header head0;
```
Comment (creator23) [inline server.cpp:37]: Good catch, I forgot to instantiate the headers and assumed they were already created.

File: main.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+int main (int argc, char* argv[]) {
+	
+	if (argc != 2) {
+	std::cerr << "Invalid call. Please call ./webserver <config_file>" <<endl;
+	return -1;
+	}
+
+
+	NginxConfigParser config_parser;
```
Comment (gabpaw987) [inline main.cpp]: Fix formatting.

File: reply.hpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#ifndef HTTP_REPLY_HPP
```
Comment (gabpaw987) [inline reply.hpp:1]: Good call to add the reply. Makes response handling much cleaner.
Also cool that the codes for following assignments are included already.

File: reply.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
+#include "reply.hpp"
+#include <string>
+
+namespace http {
+namespace server {
+
+namespace status_strings {
+
+const std::string ok =
```
Comment (gabpaw987) [inline reply.cpp:9]: Good style too!

File: server.cpp [hunk 6]
```diff
@@ -32,13 +32,12 @@ void connection::do_read() {
...
 	auto self(shared_from_this());
-	boost::asio::const_buffer a[4];
-	a[0] = boost::asio::buffer("HTTP/1.0 200 OK\r\n");
-	a[1] = boost::asio::buffer("Content-Length: " + std::to_string(reply.size()) + "\r\n");
-	a[2] = boost::asio::buffer("Content-Type: text/plain\r\n\r\n");
-	a[3] = boost::asio::buffer(reply);
 
-	boost::asio::async_write(socket_, a,
+	reply_.status = 200;
+	reply_.headers[0] = "Content-Length: " + std::to_string(reply_.content.size());
```
Comment (gabpaw987) [inline server.cpp]: Might need to be split up in name/value.

File: server.hpp [hunk 7]
```diff
@@ -12,7 +13,7 @@ class server
 public:
   server(const server&) = delete;
   server& operator=(const server&) = delete;
-  explicit server(const std::string& address, const std::string& port);
+  explicit server(const sdt::string& address, const std::string& port);
```
Comment (gabpaw987) [inline server.hpp]: Whoops, wrong namespace

File: main.cpp [hunk 8]
```diff
@@ -1,21 +1,23 @@
...
-	std::cerr << "Invalid call. Please call ./webserver <config_file>" <<endl;
+	std::cerr << "Invalid call. Please call ./webserver <config_file>" <<std::endl;
 	return -1;
 	}
 
-
-	NginxConfigParser config_parser;
-    NginxConfig config;
   	int port;
+
```
Comment (gabpaw987) [inline main.cpp:14]: Thanks for fixing the formatting! Makes it more readable.

File: reply.hpp [hunk 9]
```diff
@@ -4,11 +4,16 @@
 #include <string>
 #include <vector>
 #include <boost/asio.hpp>
-#include "header.hpp"
 
 namespace http {
 namespace server {
 
+struct header
```
Comment (gabpaw987) [inline reply.hpp:11]: 👍

File: server.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      		do_write();
+      	else
+      		do_read();
+      });
+}
+
+void connection::do_write() {
+	auto self(shared_from_this());
+	boost::asio::const_buffer a[4];
+	a[0] = boost::asio::buffer("HTTP/1.0 200 OK\r\n");
```
Comment (gabpaw987) [inline server.cpp]: Reply?

**Review by gabpaw987 — APPROVED** [review]
Looks good, will test full demo in my VM too.

File: config.cc [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+					}
+				}
+			}
+		}
+	}
+	
+	//Setting standard port if not specified.
+	portNo = 80;
+	throw InvalidConfigException("Port number missing in config file.");
+	return false;
```
Comment (nikhilshridhar) [inline config.cc:56]: Great use of throwing an exception if user is missing port number. This ensures that a valid port number is put in.

File: config.cc [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+		if(statement->tokens_[0].compare(SERVER_TOKEN) == 0 &&
+		   statement->child_block_ != nullptr)
+		{
+			for (const auto& serverStatement : statement->child_block_->statements_) 
+			{
+				if(serverStatement->tokens_[0].compare(PORT_TOKEN) == 0)
+				{
+					int portNoRead = std::stoi(serverStatement->tokens_[1]);
+					if(portNoRead > 0 && portNoRead <= 65535)
+					{
```
Comment (nikhilshridhar) [inline config.cc:37]: This is a good check to see if the port number being read is valid. Should we think about changing the range for a future assignment?

File: config_test.cc [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+TEST(ServerConfigTest, EmptyConfig) {
+  try{
+  	ServerConfig config("test_config_4");
+  	EXPECT_EQ(config.GetPortNo(), 80);
+  }
+  catch (InvalidConfigException &e)
+  {
+	std::cout << e.what();
+  }
```
Comment (nikhilshridhar) [inline config_test.cc:39]: Great unit tests! When we have to change this at all, the tests will help us determine that we don't break anything. Maybe try using a test fixture for future tests, as fixtures can make some unit tests easier to write, if they are similar.

File: config.cc [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+#include "config.h"
+
+#define SERVER_TOKEN "server"
+#define PORT_TOKEN "listen"
+
+ServerConfig::ServerConfig(const char* configFilePath) {
+  NginxConfigParser config_parser;
+  config_parser.Parse(configFilePath, &parsedConfig);
+
```
Comment (nikhilshridhar) [inline config.cc:17]: Is this object from the config_parser.h file that we saw in assignment 1?

File: config.h [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+// The in-memory representation of the entire config.
+class ServerConfig {
+ private:
+  int portNo;
+  bool ParseStatements();
+  NginxConfig parsedConfig;
+
+ public:
+  ServerConfig(const char* configFilePath);
+  int GetPortNo();
```
Comment (nikhilshridhar) [inline config.h:32]: Nice function! It makes it easier to call it in main, when the config file is passed in as an argument.

**Comment by creator23** [comment]
Seeing no other changes suggested, approving this merge.

**Pull request merged**